backend/flow_api/flow/shared/hook_oauth_links.go

```diff
@@ -0,0 +1,63 @@
+package shared
+
+import (
+	"fmt"
+	"github.com/labstack/echo/v4"
+	"github.com/teamhanko/hanko/backend/flowpilot"
+	"net/url"
+)
+
+type OAuthLinks struct {
+	Action
+}
+
+func (h OAuthLinks) Execute(c flowpilot.HookExecutionContext) error {
+	deps := h.GetDeps(c)
+
+	returnToUrl := deps.Cfg.ThirdParty.DefaultRedirectURL
+
+	referer := deps.HttpContext.Request().Header.Get("Referer")
+	if referer != "" {
+		u, err := url.Parse(referer)
+		if err != nil {
+			return err
+		}
+
+		// remove any query and fragment parts of the referer
+		u.RawQuery = ""
+		u.Fragment = ""
+		returnToUrl = u.String()
+	}
+
+	if deps.Cfg.ThirdParty.Providers.GitHub.Enabled {
+		c.AddLink(OAuthLink("github", h.generateHref(deps.HttpContext, "github", returnToUrl)))
+	}
+	if deps.Cfg.ThirdParty.Providers.Google.Enabled {
+		c.AddLink(OAuthLink("google", h.generateHref(deps.HttpContext, "google", returnToUrl)))
+	}
+	if deps.Cfg.ThirdParty.Providers.Apple.Enabled {
+		c.AddLink(OAuthLink("apple", h.generateHref(deps.HttpContext, "apple", returnToUrl)))
+	}
+
+	return nil
+}
+
+func (h OAuthLinks) generateHref(c echo.Context, provider string, returnToUrl string) string {
+	host := c.Request().Host
+	forwardedProto := c.Request().Header.Get("X-Forwarded-Proto")
+	if forwardedProto == "" {
+		// if X-Forwarded-Proto is not set, set it to http and hope that a https redirect in place
```

Comment:
```suggestion
		// If X-Forwarded-Proto is not set, set it to 'http' and hope that an HTTPS redirect is in place.
```

...but the comment within the helper function is difficult to understand without knowledge of the context in which it is used. And does it mean, that the function only works, when a redirect happened?

Reply:
Well, this function is to generate the Href of a link. And to minimise magic in the frontend part (e.g. elements) this function should also add the domain to the links, but in order to do so it needs the domain and the scheme. The domain is easy to get on, as we can get it from the `echo.Context` (`c.Request().Host`). But to know the scheme it is more difficult. Because we do not provide a way of starting Hanko with TLS, Hanko needs to be deployed behind a proxy (e.g. nginx) which does TLS termination. The proxy can send the `X-Forwarded-Proto` header which contains the scheme we need, but it can be empty. And when it is empty we have no way to know which scheme to use. 

So to answer your question, this helper function will work correctly when the proxy sends the `X-Forwarded-Proto` header or when your proxy forces an https redirect.

I can try to rewrite the comment, so that it more clear.

Reply:
I believe it would be clearer to state that you assume the proxy has set the header, rather than relying on the hope that a redirect is in place.